core/libp2p/protocol_muxer/multiselect/multiselect.cpp

```diff
@@ -5,269 +5,278 @@
-            previous_protocol = response.protocols[0];
-          }
-          break;
-        }
-        case MessageType::NA: {
-          OUTCOME_TRY(status, handleNaMsg(connection));
-          current_status = status;
-          break;
-        }
-        default:
-          return MultiselectError::INTERNAL_ERROR;
-      }
-    }
-    return finalizeNegotiation(current_status, previous_protocol);
+  void Multiselect::onWriteAckCompleted(
+      const std::shared_ptr<ConnectionState> &connection_state,
+      const Protocol &protocol) {
+    negotiationRoundFinished(connection_state, protocol);
   }
 
-  outcome::result<peer::Protocol> Multiselect::finalizeNegotiation(
-      const Status status, const peer::Protocol &protocol) const {
-    switch (status) {
-      case Status::NEGOTIATION_SUCCESS:
-        return protocol;
-      case Status::NEGOTIATION_FAIL:
-        return MultiselectError::NEGOTIATION_FAILED;
+  void Multiselect::onReadCompleted(
+      std::shared_ptr<ConnectionState> connection_state,
+      MessageManager::MultiselectMessage msg) {
+    using MessageType = MessageManager::MultiselectMessage::MessageType;
+
+    switch (msg.type) {
+      case MessageType::OPENING:
+        handleOpeningMsg(std::move(connection_state));
+        return;
+      case MessageType::PROTOCOL:
+        handleProtocolMsg(msg.protocols[0], std::move(connection_state));
+        return;
+      case MessageType::PROTOCOLS:
+        handleProtocolsMsg(msg.protocols, std::move(connection_state));
+        return;
+      case MessageType::LS:
+        handleLsMsg(std::move(connection_state));
+        return;
+      case MessageType::NA:
+        handleNaMsg(std::move(connection_state));
+        return;
       default:
-        // never should be here
-        log_->critical("achieved state, which never should be achievable");
-        return MultiselectError::INTERNAL_ERROR;
+        log_->error("type of the message, returned by the parser, is unknown");
+        MessageWriter::sendLsMsg(connection_state);
+        return;
     }
   }
 
-  outcome::result<Multiselect::Status> Multiselect::handleOpeningMsg(
-      const std::shared_ptr<basic::ReadWriteCloser> &connection,
-      Status status) const {
-    switch (status) {
-      case Status::NOTHING_SENT: {
-        // we received an opening from the other side; signalize we are here &
-        // ready
-        OUTCOME_TRY(connection->write(MessageManager::openingMsg()));
-        return Status::OPENING_SENT;
-      }
-      case Status::OPENING_SENT: {
+  void Multiselect::handleOpeningMsg(
+      std::shared_ptr<ConnectionState> connection_state) const {
+    using Status = ConnectionState::NegotiationStatus;
+
+    switch (connection_state->status_) {
+      case Status::NOTHING_SENT:
+        // we received an opening as a first message in this round; respond with
+        // an opening as well
+        MessageWriter::sendOpeningMsg(std::move(connection_state));
```

Comment:
Btw, you can write `return` before the methods returning void, so that you will not need it in the line after